cmd/commands/flatten_rbd_pvc.go

```diff
@@ -0,0 +1,147 @@
+		}
+		if pvc.DeletionTimestamp != nil {
+			logging.Fatal(fmt.Errorf("pvc %s is deleting", pvcName))
+		}
+		if pvc.Status.Phase != corev1.ClaimBound {
+			logging.Fatal(fmt.Errorf("pvc %s is not bound", pvcName))
+		}
+
+		var shouldDeleteTempImage bool
+		if pvc.Spec.DataSource != nil {
+			switch pvc.Spec.DataSource.Kind {
+			case "PersistentVolumeClaim":
+				shouldDeleteTempImage = true
+			case "VolumeSnapshot":
+			default:
+				logging.Fatal(fmt.Errorf("pvc %s is not a cloned image", pvcName))
+			}
+		}
+
+		pvName := pvc.Spec.VolumeName
+		pv, err := clientSets.Kube.CoreV1().PersistentVolumes().Get(cmd.Context(), pvName, metav1.GetOptions{})
+		if err != nil {
+			logging.Fatal(fmt.Errorf("failed to get pv %s", pvName))
+		}
+		imageName, ok := pv.Spec.CSI.VolumeAttributes["imageName"]
+		if !ok {
+			logging.Fatal(fmt.Errorf("pv %s would not be a CSI PVC", pvName))
+		}
+		poolName, ok := pv.Spec.CSI.VolumeAttributes["pool"]
+		if !ok {
+			logging.Fatal(fmt.Errorf("pv %s would not be a CSI PVC", pvName))
+		}
+
+		if !allowInUse {
+			out, err := exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rbd", []string{"-p", poolName, "status", imageName, "--format=json"}, operatorNamespace, cephClusterNamespace, false)
+			if err != nil {
+				logging.Fatal(fmt.Errorf("failed to stat %s/%s", poolName, imageName))
+			}
+			var status RBDStatusOutput
+			json.Unmarshal([]byte(out), &status)
+			if len(status.Watchers) >= 1 {
+				logging.Fatal(fmt.Errorf("flatten in-use pvc %s is not allowed. If you want to do, run with `--allow-in-use` option", pvcName))
+			}
+		}
+
+		if shouldDeleteTempImage {
+			deleteTempImage(cmd.Context(), allowInUse, poolName, imageName)
+		}
+		logging.Info("flattening the target RBD image %s/%s", poolName, imageName)
+		_, err = exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rbd", []string{"-p", poolName, "flatten", imageName}, operatorNamespace, cephClusterNamespace, false)
+		if err != nil {
+			logging.Fatal(fmt.Errorf("failed to flatten %s/%s", poolName, imageName))
+		}
+	},
+}
+
+func deleteTempImage(ctx context.Context, allowInUse bool, poolName, imageName string) {
+	tempImageName := imageName + "-temp"
+
+	out, err := exec.RunCommandInOperatorPod(ctx, clientSets, "rbd", []string{"-p", poolName, "info", "--format=json", tempImageName}, operatorNamespace, cephClusterNamespace, false)
+	if err != nil {
+		logging.Error(fmt.Errorf("failed to run `rbd info` for rbd image %s/%s", poolName, tempImageName))
+		return
+	}
+	var info RBDInfoOutput
+	json.Unmarshal([]byte(out), &info)
+	id := info.ID
+	logging.Info("removing the temporary RBD image %s/%s if exist", poolName, tempImageName)
+	_, err = exec.RunCommandInOperatorPod(ctx, clientSets, "rbd", []string{"-p", poolName, "trash", "mv", tempImageName}, operatorNamespace, cephClusterNamespace, false)
+	if err != nil {
+		logging.Fatal(fmt.Errorf("failed to move rbd image %s/%s to trash", poolName, tempImageName))
+	}
+	if id != "" {
```

Comment:
there is a corner case where CLI killed in between moving to trash and add trash remove, if someone issues the info command it fails and it will be left in trash, we can enhance it later but a pointer to be noted here.